doc/reference/index.rst

```diff
@@ -23,3 +23,214 @@ API Reference
+
+   Functions that are **sleep** may be invoked from thread context.
+
+   Functions that are **sleep** may be invoked from interrupt and
+   pre-kernel contexts if and only if invoked in **no-wait** mode.
+
+.. _api_term_no-wait:
+
+no-wait
+=======
+
+The no-wait attribute is used on a function to indicate that a parameter
+to the function can force an execution path that will not cause the
+invoking thread to sleep.
+
+.. note::
+
+   The paradigmatic case of a no-wait function is a function that takes
+   a timeout, to which `K_NO_WAIT` can be passed.  The semantics of this
+   special timeout value are to execute the function's operation as long
+   as it can be completed immediately, and to return an error code
+   rather than sleep if it cannot.
+
+   It is use of the no-wait feature that allows functions like
+   :c:func:`k_sem_take` to be invoked from ISRs, since it is not
+   permitted to sleep in interrupt context.
+
+   **NOTE** That a function has a no-wait path does not imply that taking
+   that path guarantees the function is synchronous.
+
+   Functions with this attribute may be invoked from interrupt and
+   pre-kernel contexts only when the parameter selects the no-wait path.
+
+   Functions that are **no-wait** functions are implicitly **sleep**.
+
+.. _api_term_isr-ok:
+
+isr-ok
+======
+
+The isr-ok attribute is used on a function to indicate that it can be
+called from interrupt context.  If necessary the function will use
+:c:func:`k_is_in_isr` to detect its calling context and force an
+execution path that will not cause the invoking thread to sleep.
+
+.. note:
+
+   This attribute is intended for **sleep** functions that may be
+   indirectly invoked from interrupt context with arguments that could
+   attempt to put the invoking thread to sleep, e.g. because the
+   function is not **no-wait** or the parameters do not select the
+   no-wait path.
+
+   Functions that are **isr-ok** may be always be safely invoked from
+   interrupt context, and will return an error if they were unable to
+   fulfill their behavior in that context.
+
+.. _api_term_pre-kernel-ok:
+
+pre-kernel-ok
+=============
+
+The pre-kernel-ok attribute is used on a function to indicate that it
+will take reasonable steps to ensure it is safe to invoke before all
+kernel services are started. In some cases the invocation in that
+context may return an error code.
+
+.. note::
+
+   This attribute is similar to **isr-ok** in function, but is intended
+   for use by API that is expected to be called in
+   :c:func:`DEVICE_INIT()` or :c:func:`SYS_INIT()` calls that may be
+   invoked with ``PRE_KERNEL_1`` or ``PRE_KERNEL_2`` initialization
+   levels.
+
+   Generally a function that is **pre-kernel-ok** checks
+   :c:func:`k_is_pre_kernel` when determining whether it can fulfill its
+   required behavior.  In many cases it would also check
+   :c:func:`k_is_in_isr` so it can be isr-ok as well.
```

Comment:
You're pretty consistent using bold on these names, but occasionally one slips by: ``**isr-ok**``